diff --git cuda/cublas.h cuda/cublas.h
index 970a329..b4d89ca 100644
--- cublas.h
+++ cublas.h
@@ -57,7 +57,7 @@
 #if !defined(CUBLAS_H_)
 #define CUBLAS_H_
 
-#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
 
 #ifndef CUBLASWINAPI
 #ifdef _WIN32
diff --git cuda/cudalibxt.h cuda/cudalibxt.h
index 94fcf47..fe854ee 100644
--- cudalibxt.h
+++ cudalibxt.h
@@ -53,7 +53,7 @@
 
 #ifndef _CUDA_LIB_XT_H_
 #define _CUDA_LIB_XT_H_
-#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
 
 #define CUDA_XT_DESCRIPTOR_VERSION 0x01000000 // This is added to CUDART_VERSION
 
